File: javascript/criptocurrency/Bitcoin.js
class Bitcoin {
    #price = 0;
    #observers = [];

    constructor() {
        this.#price = 0;
    }

    getPrice() {
        return this.#price;
    }

    setPrice(price) {
        if (this.#price !== price) {
            // Notify the observers
            this.notifyObservers(this.#price, price);
            this.#price = price;
        }
    }

    addObserver(observer) {
        this.#observers.push(observer);
    }

    notifyObservers(oldPrice, newPrice) {
        // Notify the observer calling the update method of all observers
        this.#observers.forEach(observer => observer.update(oldPrice, newPrice));
    }

}


module.exports = Bitcoin;

File: src/main/java/com/douglasmatosdev/criptocurrency/Bitcoin.java
package com.douglasmatosdev.criptocurrency;

import com.douglasmatosdev.service.BitcoinPriceObserver;

import java.util.ArrayList;
import java.util.List;

public class Bitcoin {
    private Double price = 0.0;
    private List<BitcoinPriceObserver> observers = new ArrayList<>();

    public Bitcoin() {
    }

    public Bitcoin(Double price) {
        this.price = price;
    }

    public Double getPrice() {
        return price;
    }

    public void setPrice(Double price) {
        if (!this.price.equals(price)) {
            this.price = price;
            // Notify the observers
            notifyObservers(this.price, price);
        }
    }

    public void addObserver(BitcoinPriceObserver observer) {
        observers.add(observer);
    }

    private void notifyObservers(Double oldPrice, Double newPrice) {
        for (BitcoinPriceObserver observer : observers) {
            // Notify the observer calling the update method of all observers
            observer.update(oldPrice, newPrice);
        }
    }
}
